lib/ecto/changeset.ex

```diff
@@ -1359,20 +1359,27 @@ defmodule Ecto.Changeset do
 
   """
   @spec validate_required(t, list | atom, Keyword.t) :: t
-  def validate_required(%{required: required, errors: errors} = changeset, fields, opts \\ []) do
+  def validate_required(%{required: required, errors: errors, changes: changes} = changeset, fields, opts \\ []) do
     message = message(opts, "can't be blank")
     fields  = List.wrap(fields)
 
-    new_errors =
+    fields_with_errors =
       for field <- fields,
           missing?(changeset, field),
           ensure_field_exists!(changeset, field),
           is_nil(errors[field]),
-          do: {field, {message, [validation: :required]}}
+          do: field
+
+    new_errors = fields_with_errors
```

Comment:
You can move this inside the case, since it is needed only when fields_with_errors is not empty. Also, don't pipe only once, you can write this as:

    new_errors = Enum.map(fields_with_error, &{&1, {message, [validation: :required]}})